File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/ManualLift.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.hardware.Gamepad;

import org.firstinspires.ftc.teamcode.hardware.LiftInternals;
import org.firstinspires.ftc.teamcode.opmodes.CycleUsingOpMode;

public class ManualLift implements LiftSubsystem {
    private final LiftInternals liftInternals;
    private final OpMode opMode;
    private final Gamepad gamepad;
    // i have NO IDEA why these can't be local variables, but this WILL break if they aren't fields
    private volatile double power;
    private volatile double lastPower;
    private volatile boolean needsUnlock;

    public ManualLift(LiftInternals liftInternals, OpMode opMode) {
        this.liftInternals = liftInternals;
        this.opMode = opMode;
        this.gamepad = opMode.gamepad1;
    }

    @Override
    public void loop() {
        // reset encoder
        if (gamepad.left_bumper && gamepad.right_bumper) liftInternals.resetEncoder();

        // rotation
        liftInternals.rotationServo.setPosition(1 - gamepad.left_trigger);

        // slide
        // TODO in kotlin this will be a Pair<Int, Boolean> and an if expression
        if (gamepad.dpad_up) {
            needsUnlock = false;
            power = (liftInternals.motor.getCurrentPosition() < LiftInternals.Position.HIGH.value + 50) ? 0 : LiftInternals.MOTOR_SCALE_FACTOR;
        } else if (gamepad.dpad_down) {
            needsUnlock = true;
            power = -LiftInternals.MOTOR_SCALE_FACTOR;
        } else {
            needsUnlock = false;
            power = 0;
        }

        if (lastPower != power) {
            liftInternals.liftExecutor.submit(() -> { // prevent float errors TODO is this comparison for float purposes needed
                if (needsUnlock) {
                    liftInternals.motor.setPower(LiftInternals.MOTOR_UNLOCK_POWER);
                    Utils.delay(100);
                    liftInternals.unlock();
                } else {
                    liftInternals.lock();
                }

                Utils.delay(100);
                liftInternals.motor.setPower(power);
            });
        }

        lastPower = power;

        // claw (closed by default)
        if (gamepad.right_trigger > 0.5) liftInternals.drop();
        else liftInternals.grab();
    }

    @Override
    public boolean canSwitch() {
        return liftInternals.motor.getCurrentPosition() > LiftInternals.Position.CAN_ROTATE.value;
    }

    @Override
    public void prepareForSwitch() {
        new Cycle((CycleUsingOpMode<?>) opMode, liftInternals, LiftInternals.Position.LOW, LiftInternals.Position.STACK_1).finish();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/ManualLift.kt
package org.firstinspires.ftc.teamcode

import org.firstinspires.ftc.teamcode.hardware.LiftInternals
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.hardware.Gamepad
import java.util.concurrent.ExecutorService

class ManualLift(private val liftInternals: LiftInternals, private val opMode: OpMode, private val cycleExecutor: ExecutorService) : LiftSubsystem {
    private val gamepad: Gamepad = opMode.gamepad1

    private var lastPower = 0.0

    override fun loop() {
        // reset encoder
        if (gamepad.left_bumper && gamepad.right_bumper) liftInternals.resetEncoder()

        // rotation
        liftInternals.rotationServo.position = 1.0 - gamepad.left_trigger

        // slide
        val (needsUnlock, power) = when {
            gamepad.dpad_up -> Pair(false,
                if (liftInternals.motor.currentPosition > LiftInternals.Position.HIGH.value + 50) 0.0
                else LiftInternals.MOTOR_SCALE_FACTOR
            )
            gamepad.dpad_down -> Pair(true, -LiftInternals.MOTOR_SCALE_FACTOR)
            else -> Pair(false, 0.0)
        }

        if (lastPower != power) {
            liftInternals.liftExecutor.submit {
                if (needsUnlock) {
                    liftInternals.motor.power = LiftInternals.MOTOR_UNLOCK_POWER
                    delay(LiftInternals.LOCK_DELAY_MS)
                    liftInternals.unlock()
                } else {
                    liftInternals.lock()
                }
                delay(100)
                liftInternals.motor.power = power
            }
        }
        lastPower = power

        // claw (closed by default)
        if (gamepad.right_trigger > 0.5) liftInternals.drop() else liftInternals.grab()
    }

    override val canSwitch get() =
        liftInternals.rotationServo.position >= 0.9 || liftInternals.motor.currentPosition > LiftInternals.Position.CAN_ROTATE.value

    override fun prepareForSwitch() {
        Cycle(
            opMode,
            cycleExecutor,
            liftInternals,
            LiftInternals.Position.LOW,
            LiftInternals.Position.STACK_1
        ).finish()
    }
}